### PR TITLE
xds-failover: adding main logic

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: xds-failover: adding main logic
Additional Description:
Adding the main logic for the xDS-failover support.
It currently supports a primary and failover sources. Initially, the manager attempts to connect to the primary twice, and if both fail, alternates between the failover and the primary until one is successfully connected to.
Once a connection is made (a response arrives) from either source, the failover manager will stick to that source.

Risk Level: low - not plumbed into any GrpcMux object yet.
Testing: Added unit test.
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
Harvey PTAL. LMK if you want more context (I've got an internal integration test the exercises this code).
/assign @htuch

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:39]: Should this be configurable?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:61]: Should these state variables be a single enum (at least the mutually exclusive ones)?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:279]: Should there be some artificial delay added to avoid rapid flip-flopping?

File: test/extensions/config_subscription/grpc/grpc_mux_failover_test.cc
Comment (htuch) [inline test/extensions/config_subscription/grpc/grpc_mux_failover_test.cc:19]: Coverage build is failing due to coverage regression, it would be good to verify all the new, somewhat intricate, code paths have full coverage.

**Review by htuch — CHANGES_REQUESTED** [review]
Approach LGTM modulo a few issues commented on.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Added a retry timer in [26aa991](https://github.com/envoyproxy/envoy/pull/34215/commits/26aa99193d553cb0e8407153157ef927e2b90f7d).
This will eventually be removed once the always trying to connect to the primary will be implemented.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:39]: There was some debate initially on whether some custom policy is needed. The main argument was not to complicate things unless that is requested.
We can always add that configuration later if needed.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:62]: In the near future the only 2 mutually exclusive ones are going to be connected_to_primary and connected_to_failover. I'm not sure if making the "connected" an enum and the "connecting" boolean flags makes the code more readable.
[Previous comment](https://github.com/envoyproxy/envoy/pull/31383#discussion_r1434205551) when a single non-exclusive flags were used, suggested using independent bools.

**Comment by adisuissa** [comment]
Assigning @soulxu as !Google maintainer reviewer.
/assign @soulxu

**Comment by adisuissa** [comment]
Unassigning @soulxu as he is OOO this week.
@wbpcode can you please take a ~Google pass at this PR?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:62]: My vote is for `connected_to` to be enum and the others boolean. Rather than readability, having structural guards against bugs here seems a net good in itself.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:39]: OK, please leave a TODO at least for future readers.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:108]: Missing coverage.

**Comment by adisuissa** [comment]
@nezdolik will you have some bandwidth to review this PR as a !Google reviewer?
Thanks in advance!

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Added some tests.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Changed to an enum for connected_to and ever_connected_to.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h:39]: Added a TODO.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:104]: do we need to explicitly set `connecting_to_primary_ = false;` here?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:82]: this method has certain degree of code duplication, wonder if it would be possible to refactor.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:35]: will this documentation be eventually exposed to users? (when code is not alpha anymore)

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Force of habit (to keep bools small, which is good when there are many instances), but I agree that it makes little sense here.
Modified as you suggested.

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: Added a note.

**Review by adisuissa — COMMENTED** [review]
Thanks for the great comments @nezdolik!

**Review by nezdolik — APPROVED** [review]
Thanks, exciting feature! There is also certain degree of code duplication across `FailoverGrpcStreamCallbacks` and `PrimaryGrpcStreamCallbacks`, but is also fine to leave it as it is if refactoring is too much effort.

**Comment by adisuissa** [comment]
> Thanks, exciting feature! There is also certain degree of code duplication across `FailoverGrpcStreamCallbacks` and `PrimaryGrpcStreamCallbacks`, but is also fine to leave it as it is if refactoring is too much effort.

Thanks!
I think that the primary and fallback will be quite different as we get to test this out and add some different behavior.
We can always refactor the code later, making them extend the same class or something.

**Comment by adisuissa** [comment]
cc @htuch for final approval and merge

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
